Add HESS DL3 DR1 validation

luca-giunti opened the pull request:
This PR addresses #2509. It is still preliminary, opened now to show the status and get feedback.

Here is a first comparison between the reference [paper](https://arxiv.org/pdf/1910.08088.pdf) results (stored [here](https://github.com/lmohrmann/hess_ost_paper_material/tree/master/spectrum_results)) and the results obtained with the new validation scripts:

**1D spectral analysis:**

- Reference:
```
amplitude: 1.62924e-11 +- 8.84563e-13 TeV-1 cm-2 s-1                 
Gamma: 2.62888 +- 0.0735089
Reference: 1.45 TeV
```

- New results
```
amplitude: 1.65e-11 +- 9.73e-13 1 / (cm2 s TeV)
index: '2.65e+00 +- 7.99e-02 '
reference: 1.45e+00 +- 0.00e+00 TeV
```

**3D analysis:**

- Reference:
```
amplitude: 1.73144e-11 +- 8.28426e-13 TeV-1 cm-2 s-1                 
Gamma: 2.69252 +- 0.0661353
Reference: 1.45 TeV
lat_0: 22.025±0.003
lon_0: 83.619±0.003
```

- New results
```
amplitude: 1.82e-11 +- 8.62e-13 1 / (cm2 s TeV)
index: '2.76e+00 +- 7.18e-02 '
reference: 1.45e+00 +- 0.00e+00 TeV
lat_0: 2.20e+01 +- 2.95e-03 deg
lon_0: 8.36e+01 +- 3.06e-03 deg
```

Comments:
1) The scripts are organized sequentially, basically as notebooks (it's the easiest way to start), but maybe defining classes or functions would be better? 
2) I implemented a sript that runs the ring background estimation, although for the crab there is no reference plot in the paper (apart from a "theta^2" plot, [which currently in gammapy is only supported at the `EventList` level](https://docs.gammapy.org/0.14/api/gammapy.data.EventList.html#gammapy.data.EventList.plot_offset2_distribution)). This script may be superfluous here, so I am not sure if we should keep it, but can be useful for the other sources
3) I computed the flux points (for both the 1D and 3D analyses), but I struggle to have the `FluxPointsEstimator` using the same energy bins as in the paper... so, for now the flux points are not really comparable with the reference
4) In the 3D analysis script, the flux points computation takes quite long, because (as described in the paper) we have to re-optimize the background norm in each energy bin. By the way, this is not possible from the HLI - I left a TODO in the script

cdeil commented:
@luca-giunti - Thanks!

> I struggle to have the FluxPointsEstimator using the same energy bins as in the paper.

@lmohrmann - what fine energy bins did you use for the analyses?
If it's not mentioned in the paper - could you please add that info in a README in the online material repo?

> Ring background maps validation

@lmohrmann - Would you be willing / do you think it's useful to share a few maps in FITS format at https://github.com/lmohrmann/hess_ost_paper_material ?

@luca-giunti - In principle OK, it's good to validate this here. My suggestion would be to focus on likelihood fits first, and to do that later, and only if reference maps become available, otherwise it's a bit pointless.

cdeil commented:
@luca-giunti - You have HTML in the `.dat` files. Please fix that. OK to put a copy here to be self-contained. But it's easiest to clone the reference file repo and then use `cp` locally for the whole reference file folder, since it's very small.

Thanks for all your work here!
I suggest you fix the `.dat` file issue and then merge this in later today.

One major concern is maintainability though. Note that Gammapy will keep changing, and you or someone else will have to update the scripts here multiple times in the coming weeks, and then the coming years. So the goal is to have as little code as possible and to avoid code duplication, just like in Gammapy code and tests. So we should try to find a way to reduce the lines of code and config files here, hopefully to 10% or 20% of what you have now. I think basically all or most source analyses are the same, just the runlist and model changes? Can we have one script with ~ 300 lines that does the whole validation?

cdeil commented:
I think introducing a single config file which has the target position as well as a few other numbers per target that are specific to each analysis (like here: https://gist.github.com/cdeil/10ebfe9a564b92cc507055293b4fb4b1) and a single `make.py` or `run.py` to run the validation for all targets would be better.
I have comments on the code here, but I think if we go through code review now and you have to manually apply the same edits in 5 Python scripts it will be very annoying.

(As I said above: I'm still +1 to merge this in today and not leave it as a long-running PR. This is a suggestion where to go next week.)

lmohrmann commented:
@cdeil I added info about the energy binning here:
https://github.com/lmohrmann/hess_ost_paper_material/blob/master/spectrum_results/README
(fine binning is 24 bins / decade)

Not sure about the maps, would be quite a bit of work to get these for all of the three tools we've used. So I don't plan to put them into the public repo. I'd suggest you guys start with the spectral results, and if you decide to also implement checks for the maps in the end, I could maybe provide just the result maps from our Gammapy analyses via "private communication".

cdeil commented:
@lmohrmann - Thanks for the quick reply!

24 bins/decade with a bin edge at 1 TeV exactly? This really matters for spectral results at high energies, and also for spectral points, which counts fall exactly in which bin. I didn't see that detail in the README. Can you add it?

> I'd suggest you guys start with the spectral results

Agree 100%, suggest to leave the 2D maps and ring to later, remove it for now.

@adonath and I several years ago already created a validation dataset where we exported HAP maps to FITS, and then checked that Gammapy gives the same result:
https://github.com/gammapy/gammapy/blob/ef80dc1ebc5b8160ee120da40af8a2f7c1246f9c/gammapy/detect/tests/test_lima.py#L30
I don't remember the details, or why Gammapy ring background estimation kept breaking and giving incorrect results in the past years a few times.
I'm not saying that old undocumented file is the way to go, just that given that there is some validation already, and that no better validation dataset is available here, it's not a good addition for now.

lmohrmann commented:
@cdeil Yes, with a bin edge at 1 TeV!

cdeil commented:
I see, the issue here is that we don't / can't run the ring in Gammapy, because we don't have the corresponding IRFs and events and exact HAP options that were used at the time:
https://github.com/gammapy/gammapy/blob/ef80dc1ebc5b8160ee120da40af8a2f7c1246f9c/gammapy/detect/tests/test_lima.py#L30

So the way to do validation here would be to take 2 runs or all runs from RXJ with some exclusion region, and run it through the same analysis with HAP and Gammapy. I guess exactly what you already did @lmohrmann . Would giving us these two FITS images and the HAP config file you used to make them a lot of work, or is it something that can be done quickly? I still don't think it's high-priority, but if it can be validated within hours, then it's nice to have.

luca-giunti commented:
I refactored a bit the code: now, there's only 1 script `run.py` of ~200 lines that runs 1D and 3D analysis looping on all the sources.

For the moment, I still implemented only 1 target (crab) and some things are still hardcoded to work for the crab, so this still requires a bit of work. But I hope it goes in the right direction.

I also removed the ring analysis.

cdeil commented:
@luca-giunti - Thanks!

You still have HTML in files like `validation/hess-dl3-dr1/crab/reference/gammapy_crab_1d_spectral_points.dat`. Please fix.

For your script: use functions or classes to structure the code.

Here's one way to structure it, by using functions:
```
def run_analysis_1d(target):
    # read config for that target
    # code to run analysis

def run_analysis_1d(target):
    # read config for that target
    # code to run analysis

def main(analyse1d=True, analyse3d=True):
    targets = ["crab", "msh1552"]
    for target in targets:
        if analyse1d:
            run_analysis_1d(target)
        if analyse3d:
            run_analysis_3d(target)

if __name__ == '__main':
    main()
```

There's other equivalent ways where you use `RunAnalysis` classes instead of functions, and the config can either be loaded on init, or passed in. That's all a matter of taste, it achieves basically the same thing: avoid code duplication, and make it possible to just run some of the analyses.

Then later the `main` function can become the command line tool entry point (like [here](https://github.com/gammapy/gammapy-benchmarks/blob/a127199a726dee9040663744390fa1a1998c9fad/benchmarks/make.py#L61)) where users can select which code to run.

cdeil commented on validation/hess-dl3-dr1/run.py:
```diff
@@ -0,0 +1,213 @@
+    flux_points = fpe.run()
+    keys = ["e_ref", "e_min", "e_max", "dnde", "dnde_errp", "dnde_errn"]
+    ascii.write(
+        flux_points.table_formatted[keys], str(PATH_RESULTS / "flux-points-3d.dat")
```
Suggest this:
```
flux_points.table_formatted[keys].write(PATH_RESULTS / "flux-points-3d.dat")
```

luca-giunti replied:
I tried, it raises the following error:
```
IORegistryError: Format could not be identified.
The available formats are:
           Format           Read Write Auto-identify Deprecated
--------------------------- ---- ----- ------------- ----------
                      ascii  Yes   Yes            No           
               ascii.aastex  Yes   Yes            No           
                ascii.basic  Yes   Yes            No           
     ascii.commented_header  Yes   Yes            No           
                  ascii.csv  Yes   Yes           Yes           
                 ascii.ecsv  Yes   Yes           Yes           
           ascii.fast_basic  Yes   Yes            No           
ascii.fast_commented_header  Yes   Yes            No           
             ascii.fast_csv  Yes   Yes            No           
       ascii.fast_no_header  Yes   Yes            No           
             ascii.fast_rdb  Yes   Yes            No           
             ascii.fast_tab  Yes   Yes            No           
          ascii.fixed_width  Yes   Yes            No           
ascii.fixed_width_no_header  Yes   Yes            No           
 ascii.fixed_width_two_line  Yes   Yes            No           
                 ascii.html  Yes   Yes           Yes           
                 ascii.ipac  Yes   Yes            No           
                ascii.latex  Yes   Yes           Yes           
            ascii.no_header  Yes   Yes            No           
                  ascii.rdb  Yes   Yes           Yes           
                  ascii.rst  Yes   Yes            No           
                  ascii.tab  Yes   Yes            No           
                       asdf  Yes   Yes           Yes           
                       fits  Yes   Yes           Yes           
                       hdf5  Yes   Yes           Yes           
                   jsviewer   No   Yes            No           
                 pandas.csv  Yes   Yes            No           
                pandas.html  Yes   Yes            No           
                pandas.json  Yes   Yes            No           
                    votable  Yes   Yes           Yes           
                     aastex  Yes   Yes            No        Yes
                        csv  Yes   Yes            No        Yes
                       html  Yes   Yes            No        Yes
                       ipac  Yes   Yes            No        Yes
                      latex  Yes   Yes            No        Yes
                        rdb  Yes   Yes            No        Yes
```

cdeil commented on validation/hess-dl3-dr1/run.py:
```diff
@@ -0,0 +1,213 @@
+            if par is not dataset.background_model.norm:
+                par.frozen = True
+
+    e_edges = (
```
Suggest we always use `MapAxis` to create energy grids instead of `np.logspace`:
https://github.com/gammapy/gammapy-benchmarks/blob/a127199a726dee9040663744390fa1a1998c9fad/benchmarks/analysis_3d.py#L29
We can try to improve the API there to make it a bit more convenient, but for now it's already a bit nicer than having to use `np.logspace(np.log10(EMIN.value), ...`

luca-giunti commented:
Thanks @cdeil for the suggestions!

> You still have HTML in files like `validation/hess-dl3-dr1/crab/reference/gammapy_crab_1d_spectral_points.dat`. Please fix.

Are you sure? I don't see them anymore, I thought I removed in my last commit...

cdeil commented:
I see HTML in `validation/hess-dl3-dr1/crab/reference/gammapy_crab_1d_powerlaw.dat` here:

https://github.com/gammapy/gammapy-benchmarks/pull/14/files#diff-41db09e011d1a6163bd2502c62a2284cR8

https://github.com/gammapy/gammapy-benchmarks/blob/e537512de466b9aa3dc23db082dc091d94e91dbd/validation/hess-dl3-dr1/crab/reference/gammapy_crab_1d_powerlaw.dat#L8

Maybe you have to push the latest version to Github?

Or maybe I'm making a mistake. If so, please share a link to the place where it's OK from here.

luca-giunti commented:
@cdeil you were right sorry. Now should be OK hopefully

cdeil commented:
> Now should be OK hopefully

👍
